sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -66,6 +68,29 @@ class SQLAppStatusListener(
         // away.
         exec.metricsValues = aggregateMetrics(exec)
         exec.write(kvstore, now)
+        if (appId != null) {
+          kvstore.write(new SQLAppStatusListenerData(appId, attemptId,
+            liveExecutions, stageMetrics))
+        }
+      }
+    }
+  }
+
+  def initialize(appId: String, attemptId: Option[String]): Unit = {
+    if (!live) {
+      this.appId = appId
+      this.attemptId = attemptId
+      try {
+        val listenerData = kvstore.read(classOf[SQLAppStatusListenerData],
+          Array(Some(appId), attemptId))
+        listenerData.liveExecutions.entrySet().asScala.foreach { entry =>
+          liveExecutions.put(entry.getKey, entry.getValue)
```

Comment:
exec.metricsValues may need to be reset to null, see comment on `kvstore.onFlush`